storage/innobase/lock/lock0lock.cc

```diff
@@ -1090,16 +1137,22 @@ static void lock_wait_wsrep(trx_t *trx)
            record-locks instead of table locks. See details
            from comment above.
         */
-        if (lock->trx != trx && lock->trx->mysql_thd &&
-            !(wsrep_thd_is_BF(lock->trx->mysql_thd, false) &&
-              wsrep_thd_order_before(lock->trx->mysql_thd, trx->mysql_thd)))
+        if (lock->trx != trx && lock->trx->mysql_thd)
         {
-          if (wsrep_thd_is_BF(lock->trx->mysql_thd, false))
+          const int has_to_wait= wsrep_BF_has_to_wait(lock, trx);
+          if (!has_to_wait)
           {
-            // There is no need to kill victim with compatible lock
-            if (!lock_has_to_wait(trx->lock.wait_lock, lock))
+            /* There is no need to kill BF victim with compatible lock. */
+            if (wsrep_thd_is_BF(lock->trx->mysql_thd, false) &&
+                !lock_has_to_wait(trx->lock.wait_lock, lock))
```

Comment:
If victim is BF, wsrep_BF_has_to_wait() may be called twice for the same victim. 
The first time, it will called few lines above. It may be called a second time when diving into lock_has_to_wait().
This might get quickly get confusing, if we move some error reporting there, as suggested in another comment.
I wonder if it would be better to change check first if the lock belongs to BF thread? Something along these lines:
```
if (wsrep_thd_is_BF(lock->trx->mysql_thd) {
  if (!lock_has_to_wait()) continue;
} else {
  if (!wsrep_BF_has_to_wait(lock, trx)) continue;
}
```

Reply:
My intent was to preserve the original code as much as possible. But if you think we can change it, then I'll give my vision of it.

First of all, `lock_wait_wsrep()` that does all BF handling has a misleading name. This function is a direct analog to `wsrep_handle_mdl_conflict()`, because its sole purpose is to BF abort conflicting lock holders. So It should be renamed to `wsrep_handle_lock_conflict()`. Then, what's its logic? To detect a conflict between wsrep transactions and then to check if this conflict is real by caling `lock_has_to_wait()`. So why don't we just call `lock_has_to_wait()` in the first place and decide BF actions based on the returned result? Considering that `lock_has_to_wait()` already calls `wsrep_BF_has_to_wait()`, I think it should do all the heavy lifting, including BF-BF checks and aborts, so that the code in `lock_wsrep_wait()` becomes like:

```
if (lock_has_to_wait(..) {
 // BF-abort
}
```
And that's it.